protostar/cairo_testing/cairo_injectable_function_runner.py

```diff
@@ -16,25 +18,60 @@ def __init__(self, hint_locals: HintLocalsDict, program: Program):
 
     async def run_cairo_function(
         self,
-        function_name: str,
+        function_identifier: OffsetOrName,
         *args: Any,
         **kwargs: Any,
     ):
         loop = asyncio.get_running_loop()
+
+        def function_runner():
+            if isinstance(function_identifier, Offset):
+                self.run_cairo_function_by_offset(function_identifier, *args, **kwargs)
+            else:
+                self.run_cairo_function_by_name(function_identifier, *args, **kwargs)
+
         await loop.run_in_executor(
             executor=None,
-            func=lambda: self.run_cairo_function_sync(function_name, *args, **kwargs),
+            func=function_runner,
         )
 
-    def run_cairo_function_sync(
+    def run_cairo_function_by_offset(
+        self,
+        offset: Offset,
+        *args: Any,
+        **kwargs: Any,
+    ):
+        with self.vm_exception_handling():
+            runner = CairoFunctionRunner(program=self._program, layout="all")
+            runner.run_from_entrypoint(
+                offset,
+                *args,
+                hint_locals=self._hint_locals,
+                static_locals={
+                    "__find_element_max_size": 2**20,
+                    "__squash_dict_max_size": 2**20,
+                    "__keccak_max_size": 2**20,
+                    "__usort_max_size": 2**20,
+                    "__chained_ec_op_max_len": 1000,
```

Comment:
Not really, since it's the original copied-over code, and i don't want to touch SN OS code since it's going to make it harder to compare and maintain

Reply:
I don't see how this would make you touch the SN OS code. I know there are places with the modified starknet code (and I agree we shouldn't touch them) but there are also places that include these constants in some tests directories (so I assume that these contents are not a part of SN OS per se and are just copied into the protostar codebase):
- `protostar/protostar/cairo_testing/cairo_injectable_function_runner.py`
- `protostar/tests/integration/cairo1_compiler_bindings/library_functions_test.py`
- `protostar/tests/integration/cairo1_compiler_bindings/tests_collector_and_compiler_test.py`

Doing what I suggested does exactly the opposite of what you mentioned - it makes it easier to compare and maintain :)

Reply:
[f43764e](https://github.com/software-mansion/protostar/pull/1522/commits/f43764e58fc7fa9ba503db7c0a5741ae3c338297)